cpp/src/Slice/Parser.cpp

```diff
@@ -2271,19 +2282,31 @@ Slice::Container::checkHasChangedMeaning(const string& name, ContainedPtr namedT
     }
     else
     {
-        // We've previously introduced the first component to the current scope, check that it has not changed meaning.
-        if (it->second->scoped() != namedThing->scoped())
+        // If both identifiers resolve to the exact same element, then their meaning didn't change.
+        if (it->second->scoped() == namedThing->scoped())
         {
-            // We don't want to issue errors for data-members or parameters.
-            // Since they can only exist within a self-contained scope, the only way for them to "change meaning"
-            // is to be redefined, which we already emit an error for elsewhere (see doesNameConflict).
-            auto isInSelfContainedScope = [](const ContainedPtr& p)
-            { return dynamic_pointer_cast<DataMember>(p) || dynamic_pointer_cast<Parameter>(p); };
+            return;
+        }
 
-            if (!isInSelfContainedScope(it->second) && !isInSelfContainedScope(namedThing))
-            {
-                unit()->error("'" + firstComponent + "' has changed meaning");
-            }
+        // If the two identifiers resolve to different elements, but those elements are in the same container,
+        // then the 'real' error is that the user redefined an element, not that an element changed meaning.
+        // This error is reported elsewhere in the code, so nothing to do here.
+        if (it->second->container() == namedThing->container())
+        {
+            return;
+        }
```

Comment:
This is the new addition to this logic.
You'll see that adding it reduced the number of unnecessary `changed meaning` errors in the `errorDetection` tests.